### PR TITLE
Add late-phase SCC disclaiming

### PR DESCRIPTION
**Pull request by cjjdespres**
We now disclaim the SCC more than once, each subsequent disclaim
occurring after a sufficient amount of time has passed in a particular
JIT phase.

Fixes: #14792
Signed-off-by: Christian Despres <despresc@ibm.com>

### CODE REVIEW

**Comment by cjjdespres** [comment]
Attn @mpirvu. I think that a trace needs to be added to the new `forceDontNeedMetadata`, but those seem to be automatically generated somehow.

File: runtime/shared_common/CacheMap.cpp [hunk 1]
```diff
@@ -229,20 +229,31 @@ SH_CacheMap::newInstance(J9JavaVM* vm, J9SharedClassConfig* sharedClassConfig, S
...
+
+/**
+ * Advise the OS to release resources associated with
+ * the metadata which have been accessed to date, regardless of
+ * whether or not this advice was issued before.
+ */
+void
+SH_CacheMap::forceDontNeedMetadata(J9VMThread* currentThread)
+{
+	SH_CompositeCacheImpl* ccToUse = _ccHead;
```
Comment (pshipton) [inline runtime/shared_common/CacheMap.cpp:250]: We can add a tracepoint to this method, similar to Trc_SHR_CM_j9shr_dontNeedMetadata

File: runtime/compiler/control/HookedByTheJit.cpp [hunk 2]
```diff
@@ -5086,6 +5089,17 @@ static void jitStateLogic(J9JITConfig * jitConfig, TR::CompilationInfo * compInf
             }
          }
       }
+   else if (!alreadyIssuedLateDisclaim
+            && (crtElapsedTime - persistentInfo->getJitStateChangeTime() >= 120000)
```
Comment (mpirvu) [inline runtime/compiler/control/HookedByTheJit.cpp:5093]: We can refine this later. Let's say that the application is under load and switches back and forth between rampup and stead-state - which I have seen happening. We may never trigger a disclaim operation.
Maybe it's better to compute the cumulative time spent in rampup/stead/deep-steady and don't care about the order of these phases or if they get interspersed with idle phases.

File: runtime/shared_common/shrinit.cpp [hunk 3]
```diff
@@ -4914,16 +4914,18 @@ adjustCacheSizes(J9PortLibrary* portlib, UDATA verboseFlags, J9SharedClassPreini
 void
 j9shr_jvmPhaseChange(J9VMThread *currentThread, UDATA phase)
 {
-	if (J9VM_PHASE_NOT_STARTUP == phase) {
-		J9JavaVM* vm = currentThread->javaVM;
+	J9JavaVM* vm = currentThread->javaVM;
 
+	if (J9VM_PHASE_NOT_STARTUP == phase) {
```
Comment (pshipton) [inline runtime/shared_common/shrinit.cpp:4919]: I think we should add a tracepoint to this function as well, that prints the new phase and the old phase.

Comment (pshipton) [reply runtime/shared_common/shrinit.cpp:4919]: Sorry, I mean that comment for jvmPhaseChange

File: runtime/vm/vmphases.c [hunk 4]
```diff
@@ -27,6 +27,7 @@
 
 void jvmPhaseChange(J9JavaVM* vm, UDATA phase) {
 	J9VMThread *currentThread = currentVMThread(vm);
+	UDATA oldPhase = vm->phase;
 	vm->phase = phase;
 	Trc_VM_VMPhases_JVMPhaseChange(phase);
```
Comment (pshipton) [inline runtime/vm/vmphases.c:32]: Pls create a new tracepoint that includes the oldPhase

**Comment by pshipton** [comment]
>  those seem to be automatically generated somehow.

Add new tracepoints to the end of j9shr.tdf for the shared classes code, and j9vm.tdf for jvmPhaseChange(). Existing tracepoint can't be modified. When an old tracepoint is no longer used, add the `Obsolete` keyword to it in the tdf file. If code tries to use an Obsolete tracepoint then it won't compile.

**Comment by pshipton** [comment]
https://github.com/eclipse-openj9/openj9/blob/master/doc/diagnostics/AddingTracepoints.md

**Review by mpirvu — COMMENTED** [review]
Somehow I dislike the idea of a so called VM phase that the VM is actually not storing, but let's worry about this later. At this point I would like to evaluate the footprint gains. So,
- Let's run AcmeAir with 80 MB SCC. Apply JMeter load and measure RSS during load
- Increase the SCC size to 120 MB (this is what we wanted to propose to Liberty team to do) and repeat the JMeter load experiment twice: with your code and without. We should see some improvement from this disclaim feature.

**Comment by hangshao0** [comment]
I am not sure if this has been measured, if we are doing more than one additional disclaim, how far are we from disclaiming the whole metadata area (area from `UPDATEPTR(_theca)` to `CADEBUGSTART(_theca)`) ? Currently `SH_CompositeCacheImpl::dontNeedMetadata()` disclaims area between min and max accessed metadata address.

**Comment by cjjdespres** [comment]
An update on the performance testing of this feature:

- Testing with the acmeair benchmark showed roughly a 3% drop in RSS after memory usage had reached a steady-state during load. In absolute terms, this was on average a 6MB drop in an RSS of 200MB.
- Testing with the daytrader 8 benchmark showed a 2-3% drop in RSS, again after memory usage had reached a steady-state during load. This corresponded to a 16MB-24MB drop in an RSS of 840MB.

@mpirvu also asked me to test if the current disclaiming affects startup time at all, and it does not seem to do so.

**Comment by cjjdespres** [comment]
The latest changes:

- The timing of the disclaim is based on total VM CPU time, not time spent in a non-idle state. Currently there is no fallback if this data is not available - the second disclaim will never occur.
- The cache access bounds continue to be updated after the first disclaim (but not the second).
- An `-XX:LateSCCDisclaimTime` option was added for testing.
- A trace was added for `forceDontNeedMetadata`. I haven't added one for old state in `jvmPhaseChanges` because I will be modifying this commit to avoid using a dummy JVM phase to control the disclaiming.

File: runtime/shared_common/CacheMap.hpp [hunk 5]
```diff
@@ -39,6 +39,8 @@
 #define CM_CACHE_CORRUPT -2
 #define CM_CACHE_STORE_PREREQ_ID_FAILED -3
 
+#define MAX_METADATA_RELEASES 2
```
Comment (hangshao0) [inline runtime/shared_common/CacheMap.hpp:42]: Please use `CM_CACHE_XXX` for the macro.

**Review by mpirvu — COMMENTED** [review]
I have a few small inline comments.
I would like to avoid that temporary phase change in vmphases.c. It's ugly and I am concerned about a concurrent thread inquiring about the VM phase and reading that temporary value. I see two options:
1. Do not store `J9VM_PHASE_LATE_SCC_DISCLAIM` into `vm->phase`; just pass `J9VM_PHASE_LATE_SCC_DISCLAIM` directly to `vm->sharedClassConfig->jvmPhaseChange(currentThread, phase);`
2. Have another SCC callback that has nothing to do with phase changes and call it directly from HookedByTheJIT.cpp

File: runtime/compiler/control/HookedByTheJit.cpp [hunk 6]
```diff
@@ -5086,6 +5087,20 @@ static void jitStateLogic(J9JITConfig * jitConfig, TR::CompilationInfo * compInf
             }
          }
       }
+   else if (lateDisclaimNeeded)
```
[the remark this answers is not in the document]
Comment (cjjdespres) [reply runtime/compiler/control/HookedByTheJit.cpp:5090]: You're right, we should. An artifact of the previous heuristic.

File: runtime/shared_common/CacheMap.cpp [hunk 7]
```diff
@@ -229,20 +229,33 @@ SH_CacheMap::newInstance(J9JavaVM* vm, J9SharedClassConfig* sharedClassConfig, S
...
+	SH_CacheMap::forceDontNeedMetadata(currentThread);
+}
+
+/**
+ * Advise the OS to release resources associated with
+ * the metadata which have been accessed to date, regardless of
+ * whether or not this advice was issued before.
+ */
+void
+SH_CacheMap::forceDontNeedMetadata(J9VMThread* currentThread)
```
[the remark this answers is not in the document]
Comment (cjjdespres) [reply runtime/shared_common/CacheMap.cpp:248]: Yes, that's possible, now that the number of disclaims is tracked.

**Comment by cjjdespres** [comment]
It does not seem to improve footprint after load. Though, now that I look at the vlog more closely, the current default has the disclaim occur at t=22745 (at least in one run) compared to the transition to NOT_STARTUP at t=225431, when the original disclaim should occur. So the current default should at least be increased. I can print out the total CPU time it takes to transition out of startup to get a better idea of what the default should be.

**Comment by mpirvu** [comment]
`NOT_STARTUP at t=225431` that doesn't look right. We should be out of start-up after 10 seconds tops.

**Comment by cjjdespres** [comment]
Could it be the container setup? Otherwise I'm still using `-Xmx1024M -Xscmx120M` in addition to the other JVM options. I could check acmeair to see if the problem persists with the JDK I built.

**Comment by mpirvu** [comment]
i don't have reasons to believe it's the container. One possibility is that you applied load immediately after start-up and the JIT could not distinguish between Liberty start-up and ramp-up. In can tell if you show me the vlog.

**Comment by cjjdespres** [comment]
That makes sense. Usually I give it a bit of time to start up, but I might have been a little too fast that time.

**Comment by cjjdespres** [comment]
Using `-XX:LateSCCDisclaimTime=1800000` (corresponding to `_lateSCCDisclaimTime=1800000000000` internally) I get an average post-load RSS of around 804MB for DT8, compared to an RSS of 832MB without a second disclaim. The disclaim tends to occur a bit after two minutes into the load with this value. (The difference is a bit lower than what was reported above because the number above is the immediate RSS drop due to the disclaim, not the post-load RSS; the immediate drop with this particular value is roughly the same as the above). I've updated the default to this value.

**Review by hangshao0 — APPROVED** [review]
The VM change LGTM.

**Comment by mpirvu** [comment]
I ran some DT8 throughput footprint experiments and I think we can use a value 300 seconds for the CPU at the point where the disclaim is issued.

Footprint values at end of run (averages of 10 runs):
80MB SCC, no disclaim ==> 779, 784 (I has 2 runs for this)
128 MB SCC, no disclaim  ==> 819 MB
128 MB SCC, disclaim after 4 min ==> 806 MB
128 MB SCC, disclaim after 5 min ==> 793, 802 MM

**Review by mpirvu — APPROVED** [review]
LGTM

**Comment by mpirvu** [comment]
@cjjdespres Please squash all commits into a single one and I'll start testing.

**Comment by mpirvu** [comment]
jenkins test sanity all jdk17

**Pull request merged**